### ISSUE TITLE
LSTM numerical gradient check

### ISSUE THREAD

**Issue by bartvm** [Element-Research/rnn#31]
Haven't been able to figure out what's going on here... With the new `LinearBias` I can get one step to give the correct gradient, but as soon as I have multiple steps weird stuff starts happening. The numerical gradients are generally much, much larger (`1e3` vs. `1e-3`). Any ideas as to what is happening?

```lua
nn = require 'nn'
require 'rnn'
require 'optim'

hiddenSize = 2
nIndex = 2
r = nn.LSTM(hiddenSize, hiddenSize)

rnn = nn.Sequential()
rnn:add(r)
rnn:add(nn.Linear(hiddenSize, nIndex))
rnn:add(nn.LogSoftMax())

criterion = nn.ClassNLLCriterion()

function f(x)
  parameters:copy(x)
  -- Do the forward prop
  rnn:zeroGradParameters()
  -- With or without fastBackward doesn't matter
  r.fastBackward = false
  local err = 0
  for i = 1, inputs:size(1) do
     local output = rnn:forward(inputs[i])
     err = err + criterion:forward(output, targets[i])
     local gradOutput = criterion:backward(output, targets[i])
     rnn:backward(inputs[i], gradOutput)
  end
  r:backwardThroughTime()
  r:forget()
  return err, grads
end

parameters, grads = rnn:getParameters()

-- This works:
-- targets = torch.Tensor{1}:resize(1, 1)
-- inputs = torch.randn(1, 2)

targets = torch.Tensor{1, 2}:resize(2, 1)
inputs = torch.randn(2, 2)
local err, dC, dC_est = optim.checkgrad(f, parameters:clone())
-- Print the exact and numerical gradients side by side
print(torch.cat(dC:view(dC:size(1), 1), dC_est:view(dC_est:size(1), 1), 2))
assert(err < 0.0001, "failed")
print("passed")
```

**Comment by bartvm** [comment]
So strangely enough I'm finding that the forward prop of the `LSTM` contains significant randomness for steps greater than one. I added `print` statements in `LSTM`'s `updateOutput` function, printing `input` and `self.output`, and some runs I get things like this:
```
in      -1.6971157419033        0.95369546810918
out     -0.0532725412628        -0.0054331652498848

in      -1.6971157419033        0.95369546810918
out     -0.025895731412077      -0.080890936582261

in      -1.6971157419033        0.95369546810918
out     0.0056107106875066      -0.14574601642533

in      -1.6971157419033        0.95369546810918
out     0.040373904069399       -0.19490861696476

in      -1.6971157419033        0.95369546810918
out     0.077159274808694       -0.22729154300181

in      -1.6971157419033        0.95369546810918
out     0.1144576942613 -0.24500315026035

in      -1.6971157419033        0.95369546810918
out     0.15065190802246        -0.25149433081713

in      -1.6971157419033        0.95369546810918
out     0.18422582030134        -0.25013744368497

in      -1.6971157419033        0.95369546810918
out     0.21395535997402        -0.24362335530009

in      -1.6971157419033        0.95369546810918
out     0.23902601951548        -0.23389435289547

in      -1.6971157419033        0.95369546810918
out     0.25905572514095        -0.22228419312076
```

**Comment by bartvm** [comment]
Found it, was actually my bad :) https://github.com/Element-Research/rnn/pull/32

**Comment by nicholas-leonard** [comment]
Awesome. Thanks.
